### ISSUE TITLE
Autoscaler goes into a CLBO if we run helm upgrade to enable it

### ISSUE THREAD

**Issue by andreas-kupries** [cloudfoundry-incubator/kubecf#973]
__Note__ that restarting the `asdatabase` pod does not seem to be a viable workaround for this ticket.

**Describe the bug**

This is not the same issue as #928, despite having the same description.
While the basic symptom of CLBO for AS pods is the same the origin is different.
In #928 the `asdatabase` pod is ok and the CLBO happens due to issues with `asactors`.
Here `asdatabase` is the troublesome pod, and `asactors` does not even complete its init containers, nor any of the other pods.

All init containers of `asdatabase` have run, but not all job containers are starting. The main `postgres` job fails. Inspection of the filesystem shared by the job containers shows that all the rendered certificate files for that job still contain their template variables, and not the expected certificates.

Copied from https://github.com/cloudfoundry-incubator/kubecf/issues/928#issuecomment-634919992 :

```
% cat /var/vcap/jobs/postgres/config/certificates/server.public_cert
((postgres_server.certificate))
```

The other cert files in the directory are the same, structurally, just a different template variable.

**To Reproduce**

This has so far only been seen in the reproduction of #928, where the `autoscaler` is enabled only after the main cluster is brought up, via a separate `helm upgrade ...` command. It is not known if it can happen when `autoscaler` is brought as part of the entire cluster.

This issue has been seen on the __GKE__ platform, for both __SA__ and __HA__ modes. It is not known if it can happen for other platforms as well. It is specifically not known if local `minikube` and/or `kind` deployments are affected.

This issue is __not deterministic__. While I have seen it often enough to call it somewhat regular it will not happen every time the AS is enabled.

While a race condition somewhere is suspected, this is not proven. While it is suspected that the double `dm` and `ig` jobs happening after the `helm upgrade ...` activating the feature are involved, this is not proven either.

**Expected behavior**

It is expected for `autoscaler` to start without issues.

**Environment**

Seen for the `kubecf-2.0.0-rc2` bundle, and for the then-master of kubecf while testing out the #928 fix.

**Additional context**

**Comment by mook-as** [comment]
I have _not_ reproduced this yet; however, I notice that when I apply the changes (i.e. `helm upgrade`), I get duplicate `ig` pods (sometimes, a `dm` and an `ig`, with the `dm` progressing to an `ig` later).  Not sure that is relevant though: template rendering happens in an init container in the `asdatabase` pod (`template-render`), going from `/var/vcap/all-releases/jobs-src/*` to `/var/vcap/jobs/*`, so it's not like it could be a race condition causing un-rendered templates to overwrite already rendered ones…

**Comment by andreas-kupries** [comment]
I saw duplicate dm/ig pods as well. Are you trying to repro on GKE ? (It was the only platform I was focused on, and never tried anything else).

**Comment by mook-as** [comment]
I'm on minikube for now (because everything is checked in, and it's easy to configure the deployment); perhaps I should try GKE as well to contrast…

**Comment by andreas-kupries** [comment]
Ah. Ok, for GKE deployment you can use catapult, all the necessary config is shown in the comments of ticket #928.

**Comment by andreas-kupries** [comment]
To be more specific, see https://github.com/cloudfoundry-incubator/kubecf/issues/928#issuecomment-634309394 . I put it there because I had to assemble the information as well.

**Comment by mook-as** [comment]
Reproduced this again, with slightly more logging (but not quite as much as I thought I had put in).  It appears that `template-render` was being run multiple times, the first few of which are missing the certs:

```
2020-06-08T17:37:58.030Z        DEBUG   manifest/cmd_render_job_tmpls.go:149    context: properties:
…
      ca: ((postgres_ca.certificate))
```

In fact, it looks like that `template-render` was never run with the secret:

```
$ grep ca: /tmp/stemcell/template-render.log | uniq
      ca: ((postgres_ca.certificate))
$ grep ca: /tmp/stemcell/template-render.log | uniq | wc -l
1
```
I'm suspicious though: why was it even running `template-render` if the inputs are missing?

**Comment by mook-as** [comment]
Looking further (at my broken cluster), I have:
secret `ig-resolved.asdatabase-v2`:
```yaml
         tls:
           ca: ((postgres_ca.certificate))
```
That's… not going to work.

I do have a `ig-resolved.asdatabase-v1` that's got the correct-looking values though:
```yaml
         tls:
           ca: &148dd933af17f036dcb496f58162b331a9fbf6ce
             -----BEGIN CERTIFICATE-----
```

Going to redeploy _again_ to see if I can catch the `ig` logs.

**Comment by andreas-kupries** [comment]
@mook-as Maybe using `stern` ? If not that I suspect you will have to ask @manno or other cf-operator people about how to capture the ig logs.

**Comment by mook-as** [comment]
As far as I can tell:
- On a fresh deployment (i.e. not the situation here), the `dm` job from quarks-operator only runs after all the variables have settled.  By the time it hits `cf-operator util variable-interpolation` all the variables are in place, so the output is a manifest with an empty `variables` section.
- On an upgrade deployment, we get _two_ `dm` jobs:
  - The initial one, where the variables are not yet in place: we have a non-empty `variables` section as output.  This leads to an `ig` job that runs with uninterpolated variables, resulting in the [`ig-resolved.asdatabase-v2`](https://github.com/cloudfoundry-incubator/kubecf/issues/973#issuecomment-640824541) as previously mentioned.
  - A _second_ `dm` job, triggered once all the variables are ready.  This one looks like the initial (non-upgrade) case, resulting in a `bpm.asdatabase-v1` manifest with an empty `variables` section.

It appears that, at least on our GKE deployments, we have a race: _sometimes_ the second job can finish before the first.  Or at least enough of a race that we can ultimately end up with a `asdatabase` pod where the input properties are unresolved, causing the error.

I have yet to figure out what triggers the two `dm` jobs.

**Comment by manno** [comment]
This is probably related to 

Are you able to gather logs when this happens? The "predicate" logging from the controllers would be especially helpful:

```
# cf-operator
kubectl get pods -A -l name=cf-operator --no-headers=true | tail -1 | read namespace name _
echo "Predicate logging from cf-operator $namespace/$name"
kubectl logs -n "$namespace" "$name" | grep "predicate passed" | grep -v recorder

# quarks-job
kubectl get pods -A -l name=quarks-job --no-headers=true | tail -1 | read namespace name _
echo "Predicate logging for pod $namespace/$name"
kubectl logs -n "$namespace" "$name" | grep "predicate passed" | grep -v recorder

```

**Comment by mook-as** [comment]
<details><summary><a href="https://github.com/cloudfoundry-incubator/kubecf/files/4788265/cf-operator-37b0da5b5d40367c226ad9d1c09cd796651ef100cb289722349fac6d22f509fb.log"><code>cf-operator-f85cfdd9f-kl7qj</code></a></summary>

```
2020-06-16T17:14:35.645Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/uaa'
2020-06-16T17:14:51.694Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/singleton-blobstore'
2020-06-16T17:17:35.499Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/api'
2020-06-16T17:17:55.599Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/routing-api'
2020-06-16T17:18:17.770Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/cc-worker'
2020-06-16T17:18:31.426Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/scheduler'
2020-06-16T17:18:50.112Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/auctioneer'
2020-06-16T17:19:39.063Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/router'
2020-06-16T17:20:15.868Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/tcp-router'
2020-06-16T17:20:16.045Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/doppler'
2020-06-16T17:22:26.559Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/diego-cell'
2020-06-16T17:22:51.537Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/log-api'
2020-06-16T17:23:17.060Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/log-cache'
2020-06-16T17:23:30.193Z        DEBUG   statefulset-rollout-reconciler  statefulset/statefulset_rollout_controller.go:43        Update predicate passed for 'scf/credhub'
2020-06-16T17:26:58.633Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-uaa-clients-app-autoscaler-secret'
2020-06-16T17:26:58.850Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-database-password'
2020-06-16T17:26:58.876Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-service-broker-password'
2020-06-16T17:26:58.880Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-scalingengine-ca'
2020-06-16T17:26:59.334Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-scalingengine-server'
2020-06-16T17:26:59.575Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-scalingengine-client'
2020-06-16T17:26:59.577Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-eventgenerator-ca'
2020-06-16T17:26:59.724Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-eventgenerator-server'
2020-06-16T17:26:59.726Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-eventgenerator-client'
2020-06-16T17:26:59.726Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-apiserver-ca'
2020-06-16T17:26:59.748Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-apiserver-server'
2020-06-16T17:26:59.749Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-apiserver-public-ca'
2020-06-16T17:27:00.205Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-apiserver-public-server'
2020-06-16T17:27:00.650Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-apiserver-client'
2020-06-16T17:27:00.688Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-servicebroker-ca'
2020-06-16T17:27:00.826Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-servicebroker-server'
2020-06-16T17:27:01.057Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-servicebroker-client'
2020-06-16T17:27:01.232Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-servicebroker-public-ca'
2020-06-16T17:27:01.645Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-servicebroker-public-server'
2020-06-16T17:27:01.952Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-scheduler-ca'
2020-06-16T17:27:02.253Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-scheduler-server'
2020-06-16T17:27:02.274Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-scheduler-client'
2020-06-16T17:27:02.539Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-metricsserver-ca'
2020-06-16T17:27:02.757Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-metricsserver-server'
2020-06-16T17:27:02.900Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-autoscaler-metricsserver-client'
2020-06-16T17:27:03.480Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-postgres-ca'
2020-06-16T17:27:03.508Z        DEBUG   quarks-secret-reconciler        quarkssecret/quarkssecret_controller.go:50      Create predicate passed for 'scf/var-postgres-server'
```
</details>

<details><summary><a href="https://github.com/cloudfoundry-incubator/kubecf/files/4788269/quarks-job-9ced5b22499c1fca157a16f627f5ad599d204cc886aaee192854aecbb830ed45.log"><code>cf-operator-quarks-job-556b8c6c7-cx55z</code></a></summary>

```
2020-06-16T17:00:28.088Z        DEBUG   errand-reconciler       quarksjob/errand_controller.go:51       Create predicate passed for 'scf/database-seeder', existing quarksJob spec.Trigger.Strategy  matches the values 'now' or 'once'
2020-06-16T17:01:13.663Z        DEBUG   errand-reconciler       quarksjob/errand_controller.go:51       Create predicate passed for 'scf/dm', existing quarksJob spec.Trigger.Strategy  matches the values 'now' or 'once'
2020-06-16T17:01:13.701Z        DEBUG   errand-reconciler       quarksjob/errand_controller.go:51       Create predicate passed for 'scf/ig', existing quarksJob spec.Trigger.Strategy  matches the values 'now' or 'once'
2020-06-16T17:01:49.065Z        DEBUG   job-reconciler  quarksjob/job_controller.go:52  Update predicate passed for 'scf/dm-a0a87c6dc6a04500', existing batchv1.Job has changed to a final state, either succeeded or failed
2020-06-16T17:02:17.368Z        DEBUG   job-reconciler  quarksjob/job_controller.go:52  Update predicate passed for 'scf/database-seeder-aabd49e3e2f7b53b', existing batchv1.Job has changed to a final state, either succeeded or failed
2020-06-16T17:09:37.607Z        DEBUG   job-reconciler  quarksjob/job_controller.go:52  Update predicate passed for 'scf/ig-29d31dda52810b2f', existing batchv1.Job has changed to a final state, either succeeded or failed
2020-06-16T17:27:03.662Z        DEBUG   errand-reconciler       quarksjob/errand_controller.go:78       Update predicate passed for 'scf/dm', a change in it´s referenced secrets have been detected
2020-06-16T17:27:27.351Z        DEBUG   job-reconciler  quarksjob/job_controller.go:52  Update predicate passed for 'scf/dm-f62ea63196b39378', existing batchv1.Job has changed to a final state, either succeeded or failed
2020-06-16T17:28:47.630Z        DEBUG   job-reconciler  quarksjob/job_controller.go:52  Update predicate passed for 'scf/dm-a1d5d51b83de89e9', existing batchv1.Job has changed to a final state, either succeeded or failed
2020-06-16T17:34:09.702Z        DEBUG   job-reconciler  quarksjob/job_controller.go:52  Update predicate passed for 'scf/ig-0d70800f32dc2d16', existing batchv1.Job has changed to a final state, either succeeded or failed
2020-06-16T17:34:40.289Z        DEBUG   job-reconciler  quarksjob/job_controller.go:52  Update predicate passed for 'scf/ig-016f0e44d11a1ff0', existing batchv1.Job has changed to a final state, either succeeded or failed
```
</details>

**Comment by mook-as** [comment]
Waiting on operator-side fix.

**Comment by mook-as** [comment]
Possibly fixed in operator [5.1.0](https://github.com/cloudfoundry-incubator/quarks-operator/releases/tag/v5.1.0); evaluating.

**Comment by mook-as** [comment]
Fixed by https://github.com/cloudfoundry-incubator/kubecf/pull/1116.
